### ISSUE TITLE
View (Render) is not Updated after State Change - No Obvious Mutations made in Reducer

### ISSUE THREAD

**Issue by Eksapsy** [reactjs/redux#2634]
I would like to submit a:
- [x] Bug
- [ ] Feature

_It most probably is my mistake, but that's why I'm here. I'm stuck and I don't know what it is._

**Redux doesn't update/re-render the components state, but the Redux' state updates as it should.**
There are no obvious mutations from the perspective I'm seeing it. I've used so far:
- `Object.assign({}, source)`
- `slice(startIndex)` // Definitely Does not work. Tried it on JSFiddle and when I changed it to Object.assign() it worked.
- ES6 **Spread Operator**
etc.

And nothing worked.
# Code (md & gists) :
## Sub-Reducer
```
export default (state = [], action) => {
  switch (action.type) {
    case ADD_COMPONENT:
      if (state[action.id]) {
        console.error('Error: Cannot add component, id already exists');
        return state;
      } else {
        return {
          ...state,
          [action.id]: action.obj
        }
      }
    case 'debug':
      console.log('debug', state[action.id]);
      return state;

    default:
      return state;
  }
};
```
**React Component** https://gist.github.com/Eksapsy/4f0b348765ba648a6c62baa249d4f756
**Actions** https://gist.github.com/Eksapsy/31c1c84e0b5975a5a0352d24af75419c

**The root reducer does not have any problems, I'm 110% of that. I'm using it the same way with this reducer as I do with my other _Working_ reducers**

_I wish I could provide a demo in JSFiddle, but I can't reproduce the bug. It's just different code. Actually, when I use some of the code in JSFiddle the code works, but it isn't the same code as in my project's_

**react-redux**: **5.0.6**",
**redux**: **3.7.2**,
**redux-thunk**: **2.2.0**,

**Comment by timdorr** [comment]
See this guide: https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/redux.md

**Comment by Eksapsy** [comment]
The truth is that I got Router in my Project. I tried injecting it with `withRouter` but still no result.

**Comment by Eksapsy** [comment]
Do I have to make another issue with the same content now or what? You closed it without considering if that's my issue or not.
And I don't see anywhere a sign that I'm using Router. So how were you so sure that this was my issue and you instantly closed it?

**Comment by markerikson** [comment]
Yeah, the Router thing was probably jumping to conclusions a bit.

However, a few other thoughts.

First, this _is_ a bug tracker, not a support forum.  You should try posting this question over on Stack Overflow or asking it in the Reactiflux chat channels.  There will be more people available who can see the question and try to help.

It's not clear from the sample code you linked or the description you gave what exactly you mean by "Redux isn't updating the component's state".

Copying values from props into an instance field like `this.reduxState` is a very odd pattern, and is probably not a good idea.  

It looks like you're actually including the component instance itself in the action, and inserting that instance into the store.  That is a bad idea - **please do not do that**.

It also looks like your reducer is defining its initial state as an array, but then trying to update it as if it were an object.  That won't work properly.

If you are trying to keep per-component state in a Redux store, [there's many existing libs that implement that capability](https://github.com/markerikson/redux-ecosystem-links/blob/master/component-state.md) - you should try using one of those instead (like `redux-ui`, `react-redux-local`, or `lean-redux`).

**Comment by Eksapsy** [comment]
Hey @markerikson. I just joined after a long day in this issue to just delete my Comment because it was too rude and I'm sorry I brought my personal problems here. :) Timdorr actually helped me very much, and I thanked him in the first place, even if this was not my problem but as you said it's not a support forum. 

Hm, your suggestions also helped much! Thanks!

Have a happy rest of your day and thank you for supporting this beautiful product, guys!
